### DISCUSSION TITLE
Have a type recipe affect the axis

### DISCUSSION THREAD

**Discussion by grero** [JuliaPlots/Makie.jl#4174] [Ideas]
I'm sure this is very obvious, but I just can't wrap my head around it. I am following the molecular simulation recipe that create a plot function for my own type. My question is how do I, through the type recipe formalism, specify e.g. axis names? There doesn't seem to be a way to get the axis object from the AbstractPlotting.plot! function that I am overriding for my type. Naively, I would want to insert something like this in my function

```julia
function AbstractPlotting.plot!(p::Plot(MyType))
  p[Axis][:names, :axisnames][] = ("x","y")
end
```

but this gives me a `MethodError` saying 
```julia
MethodError: no method matching getindex(::Combined{Any, Tuple{MyType}}, ::Type{Axis})
```

What is the proper way of overriding Axis attributes for a custom type?

**Comment by asinghvi17** [comment]
You can probably specify an `axis2d` and `axis3d` Theme in the `@recipe` theme itself.

That's how we do it in MakieThemes, here:
https://github.com/JuliaPlots/MakieThemes.jl/blob/29625720b25557dfb3fe9306022a249853257ad2/src/GGThemr/GGThemr.jl#L24-L53

**Comment by grero** [comment]
Thanks for the reply. However, that doesn't quite seem to work. This is what I tried:

```julia
struct MyType
end

function AbstractPlotting.default_theme(scene::SceneLike, ::Type{<: Plot(MyType)})
     Theme( textsize = 6,
            axis2d = Theme(frame = Theme(linewidth= 1.5,
                                           frames = ((true, false), (false, false))),
                      grid = Theme(linewidth = (0, 0)),
                      names = Theme(axisnames = ("Time", "Activity")))
         )
 end

function AbstractPlotting.plot!(p::Plot(MyType))
    x = range(0.0, stop=1.0, length=10)
    y = x.^2
    lines!(p, x,y)
end

Q = MyType()
p = plot(Q)
@test p[Axis][:frame, :frames][] == ((true, false), (false, false))
```
The test fails with the following:

```julia
Test Failed at /Users/roger/Documents/programming/julia/NeuralPlots/test/runtests.jl:29
  Expression: ((p[Axis])[:frame, :frames])[] == ((true, false), (false, false))
   Evaluated: ((false, false), (false, false)) == ((true, false), (false, false))
```

**Comment by asinghvi17** [comment]
Oh, it might be the scene attributes problem again.  There is a way to fix that but it's a tad hacky -basically you capture the relevant attributes and add them to the scene within the plotting pipeline.

**Comment by grero** [comment]
I started out following the molecular simulation tutorial, which does not use the @recipe macro directly (at least last time I checked). Also, if i understand things correctly, the @recipe macro is mainly used to define new types of plots, while my idea was to write a specialised plot function for my type, so that I can just write plot(MyType) and have it create a plot, e.g. a line plot, but formatted for my type. I think this is how I used to use the recipe framework in Plots.jl

**Comment by grero** [comment]
Also, is there code somewhere showing the hacky solution in action? : )

**Comment by asinghvi17** [comment]
Unfortunately not, that hack is made deep in the bowels of the plotting pipeline...there was a commit touching that code a little while ago.

**Comment by asinghvi17** [comment]
@SimonDanisch, sorry to keep pinging you on this, but it would be cool if this was a feature in higher-level recipes.
